File: dat-nuget/Saro.Dat/DatCryptoAesGcmNonce.cs
using System.Security.Cryptography;

namespace Saro.Dat;

public class DatCryptoAesGcmNonce : IDatCrypto, IDisposable
{
    private readonly DatCryptoAlgorithm _algorithm;
    private readonly byte[] _key;
    private bool _disposed;
    private const int NonceLen = 12;
    private const int TagLen = 16;

    private DatCryptoAesGcmNonce(DatCryptoAlgorithm algorithm, byte[] key)
    {
        _algorithm = algorithm;
        _key = key;
    }

    public static IDatCrypto FromBytes(DatCryptoAlgorithm alg, byte[] bytes)
    {
        if (bytes is null)
            throw new DatException(DatErrorCode.ConfigArgumentInvalid, "crypto key bytes are null");
        if (bytes.Length != GetKeySize(alg))
        {
            throw new DatException(DatErrorCode.KeyInvalid,
                $"{alg.ToText()} key must be {GetKeySize(alg)} bytes, got {bytes.Length}");
        }
        return new DatCryptoAesGcmNonce(alg, bytes);
    }

    public static IDatCrypto Generate(DatCryptoAlgorithm alg)
    {
        byte[] key = new byte[GetKeySize(alg)];
        RandomNumberGenerator.Fill(key);
        return new DatCryptoAesGcmNonce(alg, key);
    }

    private static int GetKeySize(DatCryptoAlgorithm alg) => alg switch
    {
        DatCryptoAlgorithm.IvAes128Gcm => 16,
        DatCryptoAlgorithm.IvAes256Gcm => 32,
        _ => throw new DatException(DatErrorCode.ConfigAlgUnsupported, $"unknown crypto algorithm: {alg}")
    };

    public byte[] Encrypt(byte[] bytes)
    {
        if (_disposed) throw new DatException(DatErrorCode.ManagerDisposed, nameof(DatCryptoAesGcmNonce));
        if (bytes is null)
            throw new DatException(DatErrorCode.ConfigArgumentInvalid, "payload to encrypt is null");
        if (bytes.Length == 0) return [];

        byte[] result = new byte[NonceLen + bytes.Length + TagLen];
        Span<byte> nonce = result.AsSpan(0, NonceLen);
        RandomNumberGenerator.Fill(nonce);

        using var aes = new AesGcm(_key, TagLen);
        try
        {
            aes.Encrypt(nonce, bytes, result.AsSpan(NonceLen, bytes.Length), result.AsSpan(NonceLen + bytes.Length, TagLen));
        }
        catch (CryptographicException e)
        {
            throw new DatException(DatErrorCode.CryptoBackend, "aes-gcm encrypt failed", e);
        }
        return result;
    }

    public byte[] Decrypt(byte[] bytes)
    {
        if (_disposed) throw new DatException(DatErrorCode.ManagerDisposed, nameof(DatCryptoAesGcmNonce));
        if (bytes is null)
            throw new DatException(DatErrorCode.ConfigArgumentInvalid, "ciphertext is null");
        if (bytes.Length == 0) return [];
        if (bytes.Length < NonceLen + TagLen)
            throw new DatException(DatErrorCode.CryptoDataInvalid,
                $"ciphertext must be at least {NonceLen + TagLen} bytes (iv + tag), got {bytes.Length}");

        byte[] plaintext = new byte[bytes.Length - NonceLen - TagLen];

        using var aes = new AesGcm(_key, TagLen);
        try
        {
            aes.Decrypt(
                bytes.AsSpan(0, NonceLen),
                bytes.AsSpan(NonceLen, plaintext.Length),
                bytes.AsSpan(bytes.Length - TagLen, TagLen),
                plaintext);
        }
        catch (AuthenticationTagMismatchException e)
        {
            throw new DatException(DatErrorCode.CryptoTagMismatch, "gcm authentication tag mismatch", e);
        }
        catch (CryptographicException e)
        {
            throw new DatException(DatErrorCode.CryptoBackend, "aes-gcm decrypt failed", e);
        }
        return plaintext;
    }

    public DatCryptoAlgorithm Algorithm() => _algorithm;

    public byte[] ToBytes()
    {
        if (_disposed) throw new DatException(DatErrorCode.ManagerDisposed, nameof(DatCryptoAesGcmNonce));
        return (byte[])_key.Clone();
    }

    public object Clone()
    {
        if (_disposed) throw new DatException(DatErrorCode.ManagerDisposed, nameof(DatCryptoAesGcmNonce));
        return new DatCryptoAesGcmNonce(_algorithm, (byte[])_key.Clone());
    }

    IDatCrypto IDatCrypto.Clone() => (IDatCrypto)Clone();

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        CryptographicOperations.ZeroMemory(_key);
    }
}

File: dat-maven/src/main/kotlin/me/saro/dat/crypto/DatCryptoAesGcmNonce.kt
package me.saro.dat.crypto

import me.saro.dat.exception.DatErrorCode
import me.saro.dat.exception.DatException
import javax.crypto.AEADBadTagException
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

class DatCryptoAesGcmNonce private constructor(
    private val algorithm: DatCryptoAlgorithm,
    private val key: SecretKeySpec,
): DatCrypto {
    companion object {
        private const val NONCE_LEN = 12
        private const val TAG_BITS = 128
        private val RANDOM = SecureRandom()
        private val CIPHER: ThreadLocal<Cipher> = ThreadLocal.withInitial { Cipher.getInstance("AES/GCM/NoPadding") }

        private fun getKeySize(algorithm: DatCryptoAlgorithm): Int {
            return when (algorithm) {
                DatCryptoAlgorithm.IV_AES128_GCM -> 16
                DatCryptoAlgorithm.IV_AES256_GCM -> 32
            }
        }

        internal fun fromBytes(algorithm: DatCryptoAlgorithm, bytes: ByteArray): DatCrypto {
            if (bytes.size != getKeySize(algorithm)) {
                throw DatException(DatErrorCode.KEY_INVALID, "$algorithm key must be ${getKeySize(algorithm)} bytes, got ${bytes.size}")
            }
            val key = SecretKeySpec(bytes, "AES")
            return DatCryptoAesGcmNonce(algorithm, key)
        }

        internal fun generate(algorithm: DatCryptoAlgorithm): DatCrypto {
            val rand = ByteArray(getKeySize(algorithm)).apply { RANDOM.nextBytes(this) }
            val key = SecretKeySpec(rand, "AES")
            return DatCryptoAesGcmNonce(algorithm, key)
        }
    }

    override fun algorithm(): DatCryptoAlgorithm {
        return algorithm
    }

    override fun toBytes(): ByteArray {
        return key.encoded
    }

    override fun encrypt(bytes: ByteArray): ByteArray {
        if (bytes.isEmpty()) {
            return ByteArray(0)
        }
        val nonce = ByteArray(NONCE_LEN).apply { RANDOM.nextBytes(this) }
        val cipher = CIPHER.get()
        return try {
            cipher.init(Cipher.ENCRYPT_MODE, key, GCMParameterSpec(TAG_BITS, nonce))
            val rv = ByteArray(NONCE_LEN + cipher.getOutputSize(bytes.size))
            System.arraycopy(nonce, 0, rv, 0, NONCE_LEN)
            cipher.doFinal(bytes, 0, bytes.size, rv, NONCE_LEN)
            rv
        } catch (e: Exception) {
            throw DatException(DatErrorCode.CRYPTO_BACKEND, "aes-gcm encrypt failed", e)
        }
    }

    override fun decrypt(bytes: ByteArray): ByteArray {
        if (bytes.isEmpty()) {
            return ByteArray(0)
        }
        if (bytes.size <= NONCE_LEN) {
            throw DatException(DatErrorCode.CRYPTO_DATA_INVALID, "ciphertext is shorter than the 12-byte iv")
        }
        val cipher = CIPHER.get()
        return try {
            cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(TAG_BITS, bytes, 0, NONCE_LEN))
            cipher.doFinal(bytes, NONCE_LEN, bytes.size - NONCE_LEN)
        } catch (e: AEADBadTagException) {
            throw DatException(DatErrorCode.CRYPTO_TAG_MISMATCH, "gcm authentication tag mismatch", e)
        } catch (e: DatException) {
            throw e
        } catch (e: Exception) {
            throw DatException(DatErrorCode.CRYPTO_BACKEND, "aes-gcm decrypt failed", e)
        }
    }

    override fun clone(): DatCrypto {
        return fromBytes(algorithm, toBytes())
    }
}